snapshot/snapshotfs/upload.go

```diff
@@ -255,29 +255,52 @@ func (u *Uploader) uploadDir(ctx context.Context, rootDir fs.Directory, policyTr
 	return de, err
 }
 
-func (u *Uploader) foreachEntryUnlessCancelled(relativePath string, entries fs.Entries, cb func(entry fs.Entry, entryRelativePath string) error) error {
-	for _, entry := range entries {
-		if u.IsCancelled() {
-			return errCancelled
-		}
+func (u *Uploader) foreachEntryUnlessCancelled(ctx context.Context, parallel int, relativePath string, entries fs.Entries, cb func(ctx context.Context, index int, entry fs.Entry, entryRelativePath string) error) error {
+	type entryWithIndex struct {
+		entry fs.Entry
+		index int
+	}
 
-		entryRelativePath := relativePath + "/" + entry.Name()
+	ch := make(chan entryWithIndex)
+	eg, ctx := errgroup.WithContext(ctx)
 
-		if err := cb(entry, entryRelativePath); err != nil {
-			return err
+	// one goroutine to pump entries into channel until ctx is closed.
+	eg.Go(func() error {
+		defer close(ch)
+
+		for i, e := range entries {
+			select {
+			case ch <- entryWithIndex{e, i}: // sent to channel
+			case <-ctx.Done(): // context closed
+				return nil
+			}
 		}
-	}
+		return nil
+	})
 
-	return nil
-}
+	// launch N workers in parallel
```

Comment:
Should `parallel` be capped to the number of entries as well for directories with just a couple of entries?

Reply:
done